peripheral/stm32g4/uart.cpp

```diff
@@ -0,0 +1,224 @@
+}
+
+void Uart::initGpio()
+{
+  // Configure GPIO alternate functions
+  setGpioAlternateFunction(
+    init_struct_.gpioPort,
+    init_struct_.gpioPinTx,
+    init_struct_.gpioAlternateFunction
+  );
+
+  setGpioAlternateFunction(
+    init_struct_.gpioPort,
+    init_struct_.gpioPinRx,
+    init_struct_.gpioAlternateFunction
+  );
+
+  // Configure output speed to "Very high speed" for MISO
+  init_struct_.gpioPort->OSPEEDR |= (0x03 << (init_struct_.gpioPinTx * 2));
+}
+
+void Uart::initDma()
+{
+  // UART Rx Dma channel
+  init_struct_.rxDmaChannel->CPAR = (uint32_t)&init_struct_.usart->RDR;
+  init_struct_.rxDmaChannel->CNDTR = 0;
+  init_struct_.rxDmaChannel->CCR = DMA_CCR_MINC | DMA_CCR_TCIE;
+
+  init_struct_.rxDmaMuxChannel->CCR = init_struct_.rxDmaMuxId;
+
+  // UART Tx Dma channel
+  init_struct_.txDmaChannel->CPAR = (uint32_t)&init_struct_.usart->TDR;
+  init_struct_.txDmaChannel->CNDTR = 0;
+  init_struct_.txDmaChannel->CCR = DMA_CCR_DIR | DMA_CCR_MINC | DMA_CCR_TCIE;
+
+  init_struct_.txDmaMuxChannel->CCR = init_struct_.txDmaMuxId;
+
+  NVIC_SetPriority(
+    init_struct_.rxDmaIrqN,
+    NVIC_EncodePriority(NVIC_GetPriorityGrouping(), init_struct_.rxDmaIrqPriority, 0)
+  );
+
+  NVIC_SetPriority(
+    init_struct_.txDmaIrqN,
+    NVIC_EncodePriority(NVIC_GetPriorityGrouping(), init_struct_.txDmaIrqPriority, 0)
+  );
+
+  NVIC_EnableIRQ(init_struct_.rxDmaIrqN);
+  NVIC_EnableIRQ(init_struct_.txDmaIrqN);
+
+}
+
+void Uart::initUart()
+{
+  init_struct_.usart->BRR = init_struct_.brrValue;
+  init_struct_.usart->CR3 = USART_CR3_DMAT | USART_CR3_DMAR;
+  init_struct_.usart->CR1 = USART_CR1_PCE | USART_CR1_M0 | USART_CR1_FIFOEN | USART_CR1_TE | USART_CR1_RE | USART_CR1_UE;
+}
+
+void Uart::init()
+{
+  FIGURE_ASSERT(is_initialized_ == false, "Already initialized");
+
+  initClock();
+  initGpio();
+  initDma();
+  initUart();
+
+  instance = this;
+
+  is_initialized_ = true;
+}
+
+void Uart::reset()
+{
+  // Reset UART
+  *init_struct_.uartRccResetRegister |= (1 << init_struct_.uartRccResetBit);
+  *init_struct_.uartRccResetRegister &= ~(1 << init_struct_.uartRccResetBit);
+  initUart();
+}
+
+void Uart::startTransaction(BufferDescriptor descriptor)
+{
+  FIGURE_ASSERT(!(descriptor.rxBuffer != NULL && descriptor.txBuffer != NULL)); // Tx and Rx at the same time is not supported by UART driver
+
+  transaction_counter_++;
+
+  // Disable the DMA
+  init_struct_.rxDmaChannel->CCR &= ~DMA_CCR_EN;
+  init_struct_.txDmaChannel->CCR &= ~DMA_CCR_EN;
+
+  // Configure rxBuffer
+  if(descriptor.rxBuffer != NULL)
+  {
+    init_struct_.rxDmaChannel->CNDTR = descriptor.length;
+    init_struct_.rxDmaChannel->CMAR = (uint32_t)descriptor.rxBuffer;
+    init_struct_.rxDmaChannel->CCR |= DMA_CCR_EN;
+  }
+
+  // Configure txBuffer
+  if(descriptor.txBuffer != NULL)
+  {
+    init_struct_.txDmaChannel->CNDTR = descriptor.length;
+    init_struct_.txDmaChannel->CMAR = (uint32_t)descriptor.txBuffer;
```

Comment:
make  rxBuffer and txBuffer volatile to prevent them getting optimized out.

Reply:
fixed.